### PR TITLE
Add a function to merge trivia from a node into an existing trivia

### PR DESCRIPTION
**Pull request by bnbarham**
Generally when a node is removed, we need to add its trivia to an
existing node. Add a small helper function to aid in this case.

### CODE REVIEW

**Comment by bnbarham** [comment]
@swift-ci please test

File: Sources/SwiftSyntax/generated/TriviaPieces.swift
Comment (ahoppen) [inline Sources/SwiftSyntax/generated/TriviaPieces.swift:1]: You need to update CMakeLists.txt and group.json. I always forget as well

File: Sources/SwiftSyntax/Trivia.swift [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+  /// Creates a new `Trivia` by appending the provided `TriviaPiece` to the end.
+  public func appending(_ piece: TriviaPiece) -> Trivia {
+    var copy = pieces
+    copy.append(piece)
+    return Trivia(pieces: copy)
+  }
+
+  /// Creates a new `Trivia` by appending the leading and trailing `Trivia`
+  /// from the given to the end.
+  public func appending<T: SyntaxProtocol>(_ node: T) -> Trivia {
```
[the remark this answers is not in the document]
Comment (bnbarham) [reply Sources/SwiftSyntax/Trivia.swift:47]: I like that, thanks.

**Comment by bnbarham** [comment]
This one actually depends on https://github.com/apple/swift-syntax/pull/1477 so I'll wait to re-run tests until it's in.

File: Sources/SwiftSyntax/Trivia.swift [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+  public var isEmpty: Bool {
+    pieces.isEmpty
+  }
+
+  /// Creates a new `Trivia` by appending the provided `TriviaPiece` to the end.
+  public func appending(_ piece: TriviaPiece) -> Trivia {
+    var copy = pieces
+    copy.append(piece)
+    return Trivia(pieces: copy)
+  }
```
[the remark this answers is not in the document]
Comment (ahoppen) [reply Sources/SwiftSyntax/Trivia.swift:43]: OK, so let’s get rid of the `+` overload now 🗑️

[the remark this answers is not in the document]
Comment (bnbarham) [reply Sources/SwiftSyntax/Trivia.swift:47]: This is used... in quite a few places. I'd rather get this in for now, so another PR it is 😅

File: Sources/SwiftSyntax/generated/raw/RawSyntaxValidation.swift [hunk 3]
```diff
@@ -227,7 +227,12 @@ func validateLayout(layout: RawSyntaxBuffer, as kind: SyntaxKind) {
   case .accessPathComponent:
     assert(layout.count == 5)
     assertNoError(kind, 0, verify(layout[0], as: RawUnexpectedNodesSyntax?.self))
-    assertNoError(kind, 1, verify(layout[1], as: RawTokenSyntax.self, tokenChoices: [.tokenKind(.identifier)]))
```
Comment (bnbarham) [inline Sources/SwiftSyntax/generated/raw/RawSyntaxValidation.swift:230]: Seems like we didn't regenerate in https://github.com/apple/swift-syntax/commit/aea4442aba9f0393de87493b13df32962d95f37a?

**Comment by bnbarham** [comment]
@swift-ci please test

**Comment by bnbarham** [comment]
@swift-ci please test

File: Sources/SwiftSyntax/Trivia.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (ahoppen) [reply Sources/SwiftSyntax/Trivia.swift:47]: 👍🏽 Maybe we can deprecate it in a follow-up PR 🤔

**Comment by bnbarham** [comment]
@swift-ci please test

File: Sources/SwiftSyntax/Trivia.swift [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+  /// Creates a new `Trivia` by appending the given trivia to the end.
+  public func appending(_ trivia: Trivia) -> Trivia {
+    var copy = pieces
+    copy.append(contentsOf: trivia.pieces)
+    return Trivia(pieces: copy)
+  }
+
+  /// Creates a new `Trivia` by merging in the given trivia. Drops any leading
+  /// trivia that matches the trailing trivia of `self`.
+  public func merging(_ trivia: Trivia) -> Trivia {
```
Comment (ahoppen) [inline Sources/SwiftSyntax/Trivia.swift:58]:
```suggestion
  /// Creates a new `Trivia` by merging in the given trivia. Only includes one copy of a common prefix of `self` and `trivia`.
  public func merging(_ trivia: Trivia) -> Trivia {
```

File: Sources/SwiftSyntax/Trivia.swift [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+      if lhs.suffix(infixLength) == rhs.suffix(infixLength) {
+        return lhs.appending(Trivia(pieces: Array(rhs.dropFirst(infixLength))))
+      }
+    }
+    return lhs.appending(rhs)
+  }
+
+  /// Creates a new `Trivia` by merging the leading and trailing `Trivia`
+  /// of `triviaOf` into the end of `self`. Drops any trivia that matches the
+  /// the trailing trivia of `self`.
```
Comment (ahoppen) [inline Sources/SwiftSyntax/Trivia.swift:71]:
```suggestion
  /// Creates a new `Trivia` by merging the leading and trailing `Trivia`
  /// of `triviaOf` into the end of `self`. Only includes one copy of any common prefixes.
```

**Review by ahoppen — APPROVED** [review]

**Comment by bnbarham** [comment]
@swift-ci please test

**Comment by bnbarham** [comment]
@swift-ci please test

**Comment by bnbarham** [comment]
@swift-ci please test Windows platform

**Comment by bnbarham** [comment]
@swift-ci please test Windows platform

**Pull request merged**